### PR TITLE
Feature: Market browser

### PR DESCRIPTION
**Pull request by recursivetree**
Adds a market browser using the new market order data. This started so we have a use for the market orders in the core, but it is regardless a useful tool.

Bildschirmfoto 2023-03-05 um 13 34 47

### CODE REVIEW

**Comment by recursivetree** [comment]
of course, this can only be merged together with the pr to eveapi

File: src/Http/Controllers/Tools/MarketController.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * Class MarketController.
+ *
+ * @package Seat\Web\Http\Controllers\Tools
+ */
+class MarketController extends Controller
+{
+    const DEFAULT_ITEM = 4358;
+
+    public function browser(MarketOrderDataTable $dataTable) {
+        $type_id = intval(request()->input('type_id', 0));
```
Comment (warlof) [inline src/Http/Controllers/Tools/MarketController.php:41]:
```suggestion
        $type_id = intval(request()->input('type_id', self::DEFAULT_ITEM));
```

File: src/Http/DataTables/Scopes/Filters/SellOrderScope.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+     * @var bool
+     */
+    private $sell_orders;
+
+    /**
+     * TypeScope constructor.
+     *
+     * @param  bool  $sell_orders
+     */
+    public function __construct(int $sell_orders) {
```
Comment (warlof) [inline src/Http/DataTables/Scopes/Filters/SellOrderScope.php:44]:
```suggestion
    public function __construct(bool $sell_orders) {
```

**Comment by recursivetree** [comment]
thanks for the things from the code review!

File: src/Http/DataTables/Tools/MarketOrderDataTable.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    {
+        return MarketOrder::with('type:typeName', 'station:station_id,name', 'solar_system:system_id,name,security', 'structure:structure_id,name');
+    }
+
+    /**
+     * @return array
+     */
+    public function getColumns() {
+        return [
+            ['data' => 'system_id', 'title' => 'System'],
```
Comment (warlof) [inline src/Http/DataTables/Tools/MarketOrderDataTable.php:94]:
```suggestion
            ['data' => 'solar_system.name', 'title' => 'System'],
```

you're displaying the system name. Current syntax will prevent proper searchability and sorting

File: src/Http/DataTables/Tools/MarketOrderDataTable.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            ['data' => 'volume_remaining', 'title' => 'Quantity'],
+            ['data' => 'price', 'title' => 'Price'],
+            ['data' => 'location_id', 'title'=>'Location'],
```
Comment (warlof) [inline src/Http/DataTables/Tools/MarketOrderDataTable.php:97]: that one would need a revamp once the upwell/station/outpost PR will be sorted (eveseat/eveapi#343)

File: src/resources/views/partials/location.blade.php [hunk 5]
```diff
@@ -2,19 +2,12 @@
   {{ trans('web::seat.unknown') }}
 @else
   @if(! is_null($location->solar_system))
-  <a href="//evemaps.dotlan.net/system/{{ $location->solar_system->name }}" target="_blank">
-    <span class="fas fa-map-marker-alt"></span>
-  </a>
-  @endif
-  @if(!is_null($location->structure))
+    @include("web::partials.system",["system"=>$location->solar_system->name,"security"=>$location->solar_system->security])
```
Comment (warlof) [inline src/resources/views/partials/location.blade.php:5]:
```suggestion
    @include('web::partials.system', ['system' => $location->solar_system->name, 'security' => $location->solar_system->security])
```

use simple quote instead double ones, double quote is limisted to html usage

File: src/Http/DataTables/Tools/MarketOrderDataTable.php [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+class MarketOrderDataTable extends DataTable
+{
+    /**
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function ajax(): JsonResponse
+    {
+        return datatables()
+            ->eloquent($this->applyScopes($this->query()))
+            ->editColumn('system_id', function ($row) {
```
Comment (warlof) [inline src/Http/DataTables/Tools/MarketOrderDataTable.php:44]:
```suggestion
            ->editColumn('solar_system.name', function ($row) {
```

**Review by warlof — COMMENTED** [review]
Just spotted an issue with search and sort usage based on system field.

Except that, I think we have to sort the new structure job in order to safely relay on it for the location field.

File: src/resources/views/partials/location.blade.php (hunk 5 above)
Comment (recursivetree) [reply src/resources/views/partials/location.blade.php:5]: I know seat uses single quotes, but I'm just used to using double quotes

File: src/Http/DataTables/Tools/MarketOrderDataTable.php (hunk 4 above)
Comment (recursivetree) [reply src/Http/DataTables/Tools/MarketOrderDataTable.php:97]: Shouldn't need it. The PR just changes the way it is loaded, but nothing about how it is stored in the db

**Comment by recursivetree** [comment]
Can be merged now, unless new issue arise

**Pull request merged**